worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,334 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package eventqueue
+
+import (
+	"sync/atomic"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/worker/v3/catacomb"
+)
+
+// Logger represents the logging methods called.
+type Logger interface {
+	Infof(message string, args ...interface{})
+	Tracef(message string, args ...interface{})
+	IsTraceEnabled() bool
+}
+
+// Stream represents a way to get change events.
+type Stream interface {
+	// Changes returns a channel for a given namespace (database).
+	Changes() <-chan changestream.ChangeEvent
+}
+
+// subscription represents a subscriber in the event queue. It holds a tomb, so
+// that we can tie the lifecycle of a subscription to the event queue.
+type subscription struct {
+	tomb tomb.Tomb
+	id   uint64
+
+	topics map[string]struct{}
+
+	changes chan changestream.ChangeEvent
+	done    <-chan struct{}
+
+	unsubscriber func()
```

Comment:
Would prefer to call this `unsubscribe` or `unsubscribeFunc`.
When I saw it below in the `Unsubscribe()` method, I thought it was a typo.